row_transform/row_transform.tex

```diff
@@ -0,0 +1,190 @@
+
+\begin{theorem}
+    For every setting of the input parameters \texttt{(atom\_input\_domain, atom\_output\_domain, atom\_function)} to \texttt{make\_row\_by\_row} such that the given preconditions
+    hold, the transformation returned by \texttt{make\_row\_by\_row} has the following properties:
+    \begin{enumerate}
+        \item \textup{(Appropriate output domain).} For every element $v$ in \texttt{input\_domain}, $\function(v)$ is in \texttt{output\_domain}. 
+        
+        \item \textup{(Domain-metric compatibility).} The domain \texttt{input\_domain} matches one of the possible domains listed in the definition of \texttt{input\_metric}, and likewise \texttt{output\_domain} matches one of the possible domains listed in the definition of \texttt{output\_metric}.
+        
+        \item \textup{(Stability guarantee).} For every pair of elements $v, w$ in \texttt{input\_domain} and for every pair $(\din, \dout)$, where $\din$ is of the associated type for \texttt{input\_metric} and $\dout$ is the associated type for \texttt{output\_metric}, if $v,w$ are $\din$-close under \texttt{input\_metric} and $\Relation(\din, \dout) = \True$, then $\function(v), \function(w)$ are $\dout$-close under \texttt{output\_metric}.
+    \end{enumerate}
+\end{theorem}
+\begin{proof}
+\begin{enumerate}
+\item \textbf{(Appropriate output domain).} In the case of \texttt{make\_row\_by\_row}, this corresponds to showing that for every vector $v$ of elements of type \texttt{DIA}, \texttt{function}$(v)$ is a vector of elements of type \texttt{DOA}. 
+
+The $\function(v)$ has type \texttt{Vec[DOA]} follows from the assumption that element $v$ is in \texttt{input\_domain} and from the type signature of \texttt{function} in line~\ref{line:fn} of the pseudocode (Section~\ref{sec:pseudocode}), which takes in an element of type \texttt{Vec(DIA)} and returns an element of type \texttt{Vec[DOA]}. If the Rust code compiles correctly, then the type correctness follows from the definition of the type signature enforced by Rust. Otherwise, the code raises a compile time error for incorrect function input type or output type. 
+
+The type signature is not sufficient because the function output type may not necessarily be contained in output domain. This can happen if output type contains a null value. Since \texttt{atom\_function} has a function trait that maps elements of \texttt{&DIA::Carrier} to \texttt{DOA::Carrier}, and the Carrier of a domain may contain \texttt{Null} values, the output type may contain a \texttt{null} value, which is outside of the output domain of \texttt{VectorDomain(DOA)}.
+
+Note that pure functions must always return a value and the precondition guarantees that \texttt{atom\_function} must be a pure function. Therefore the output type of \texttt{function} cannot contain null values. Therefore the appropriate domain property is satisfied output type is contained within the output domain.
```

Comment:
Incorporated Mike's feedback that output type may contain null value. Would like feedback on the following
1. Am I correct on the reason why output type may contain null value? (Is it because carrier may contain null values)?
2. How do we know that the output domain `VectorDomain(DOA)` must contain no null values? 
3. Is the reason why the type signature issue fixed (that the output type is not a null value) because the atom_function is a pure function?